main/src/ca/uwaterloo/flix/language/phase/Kinder.scala

```diff
@@ -0,0 +1,1077 @@
+      for {
+        specKinds <- getKindsFromSpec(spec0, root)
+        kinds <- kenv0 ++ specKinds
+        spec <- ascribeSpec(spec0, kinds, root)
+        exp <- Validation.traverse(exp0)(ascribeExpression(_, kinds, root))
+      } yield KindedAst.Sig(sym, spec, exp.headOption)
+  }
+
+  /**
+    * Performs kinding on the given spec under the given kind environment.
+    */
+  private def ascribeSpec(spec0: ResolvedAst.Spec, kenv: KindEnv, root: ResolvedAst.Root)(implicit flix: Flix): Validation[KindedAst.Spec, KindError] = spec0 match {
+    case ResolvedAst.Spec(doc, ann0, mod, tparams0, fparams0, sc0, tpe0, eff0, loc) =>
+      val annVal = Validation.traverse(ann0)(ascribeAnnotation(_, kenv, root))
+      val tparamsVal = Validation.traverse(tparams0.tparams)(ascribeTparam(_, kenv))
+      val fparamsVal = Validation.traverse(fparams0)(ascribeFormalParam(_, kenv, root))
+      val tpeVal = ascribeType(tpe0, KindMatch.subKindOf(Kind.Star), kenv, root)
+      val effVal = ascribeType(eff0, KindMatch.subKindOf(Kind.Bool), kenv, root)
+      val scVal = ascribeScheme(sc0, kenv, root)
+      for {
+        result <- Validation.sequenceT(annVal, tparamsVal, fparamsVal, tpeVal, effVal)
+        (ann, tparams, fparams, tpe, eff) = result
+        sc <- scVal // ascribe the scheme separately
+      } yield KindedAst.Spec(doc, ann, mod, tparams, fparams, sc, tpe, eff, loc)
+  }
+
+  /**
+    * Performs kinding on the given formal param under the given kind environment.
+    */
+  private def ascribeFormalParam(fparam0: ResolvedAst.FormalParam, kinds: KindEnv, root: ResolvedAst.Root)(implicit flix: Flix): Validation[KindedAst.FormalParam, KindError] = fparam0 match {
+    case ResolvedAst.FormalParam(sym, mod, tpe0, loc) =>
+      mapN(ascribeType(tpe0, KindMatch.subKindOf(Kind.Star), kinds, root)) {
+        tpe => KindedAst.FormalParam(sym, mod, tpe, loc)
+      }
+  }
+
+  /**
+    * Performs kinding on the given annotation under the given kind environment.
+    */
+  private def ascribeAnnotation(ann: ResolvedAst.Annotation, kinds: KindEnv, root: ResolvedAst.Root)(implicit flix: Flix): Validation[KindedAst.Annotation, KindError] = ann match {
+    case ResolvedAst.Annotation(name, exps0, loc) =>
+      mapN(Validation.traverse(exps0)(ascribeExpression(_, kinds, root))) {
+        exps => KindedAst.Annotation(name, exps, loc)
+      }
+  }
+
+  /**
+    * Performs kinding on the given expression under the given kind environment.
+    */
+  private def ascribeExpression(exp00: ResolvedAst.Expression, kenv: KindEnv, root: ResolvedAst.Root)(implicit flix: Flix): Validation[KindedAst.Expression, KindError] = exp00 match {
+    case ResolvedAst.Expression.Wild(loc) => KindedAst.Expression.Wild(Type.freshVar(Kind.Star), loc).toSuccess
+    case ResolvedAst.Expression.Var(sym, tpe0, loc) =>
+      mapN(ascribeType(tpe0, KindMatch.subKindOf(Kind.Star), kenv, root)) {
+        tpe => KindedAst.Expression.Var(sym, tpe, loc)
```

Comment:
So, do we have some tests like:

```
def foo(): Int32 & ef = 
  let x: Option[ef] = ...
```

Or similar? For the Var case and for Let/Ascribe?